Add gas tests for sierra-gas tracked resource

THenry14 opened the pull request:
Related to #2977

This PR stack aims to add support for new resource tracking (sierra-gas) and new gas representation (GasVector triplet instead of single l1 gas value).

-- (https://github.com/foundry-rs/starknet-foundry/pull/3049) Add sierra gas tracking
-- (https://github.com/foundry-rs/starknet-foundry/pull/3050) Run already existing tests with CairoSteps set explicitly
-- (https://github.com/foundry-rs/starknet-foundry/pull/3051) Add support for sierra gas in `--detailed-resources`
-- (https://github.com/foundry-rs/starknet-foundry/pull/3052) Add support for sierra gas in `--save-trace-data`
-- (https://github.com/foundry-rs/starknet-foundry/pull/3053) Add sierra version assertion
-- (https://github.com/foundry-rs/starknet-foundry/pull/3054) Introduce new gas representation
--> (This PR) Add gas tests for sierra-gas tracking
-- (https://github.com/foundry-rs/starknet-foundry/pull/3056) Update docs with sierra-gas related changes

THenry14 committed:
[skip ci] Add gas tests for sierra-gas tracked resource

ddoktorski reviewed (changes requested):
Even with Scarb `2.11.1` some of these tests already fail. Based on our experience so far, such tests seem unmaintainable in the long run. I don't have a specific idea on how to test gas related features, but could we at least assert all values with some acceptable margin? 🙏

cc @cptartur

cptartur commented on crates/forge/tests/integration/gas.rs:
```diff
@@ -1002,3 +1002,574 @@ fn events_contract_cost_cairo_steps() {
+    // n = 1 (unique contracts updated - in this case it's the new contract address)
+    // ( l + n * 2 ) * felt_size_in_bytes(32) = 96 (total l1 data cost)
+    //
+    // 20000 = cost of 2 keccak syscall (because 2 * 100 * 100) (from constructor)
```
Maybe we add some units to it, so it's more obvious: keccak is 100 steps, and each step is 100 sierra gas. 

And that the price of keccak is from versioned constants.

It will be much easier to maintain this in the future

THenry14 replied:
All this information is already provided 🤔 I mean, I see your point, but there are things in the top comment for a reason - to not repeat all the same stuff over and over. I have ammended the top comment with information that:
```
Costs of syscalls (if provided) are taken from versioned_constants (blockifier)
```

to be super-duper clear, but really I find this info redundant at this point tbh. However, you're the boss here, and if you insist I can add it ofc 🙏

cptartur commented on crates/forge/tests/integration/gas.rs:
```diff
@@ -1002,3 +1002,574 @@ fn events_contract_cost_cairo_steps() {
+    // 96 = gas cost of onchain data (see `deploy_syscall_cost_sierra_gas` test)
+    // 20000 = cost of 2 keccak syscall (see `deploy_syscall_cost_sierra_gas` test)
+    // 142810 = cost of 1 deploy syscall (see `deploy_syscall_cost_sierra_gas` test)
+    // 487334 = reported consumed sierra gas
```
So this is just from code execution that doesn't directly translate to anything specific (like syscalls etc.)?

cptartur commented on crates/forge/tests/integration/gas.rs:
```diff
@@ -1002,3 +1002,574 @@ fn events_contract_cost_cairo_steps() {
+    let result = run_test_case(&test, ForgeTrackedResource::SierraGas);
+
+    assert_passed(&result);
+    // 29524 = gas cost of l2 -> l1 message
```
I assume this was just taken from what was reported back from the gas calculation code?

THenry14 replied:
Yup

cptartur replied:
Maybe it makes sense to link this https://github.com/foundry-rs/starknet-foundry/issues/2960 issue here and in related?

cptartur commented on crates/forge/tests/integration/gas.rs:
```diff
@@ -1002,3 +1002,574 @@ fn events_contract_cost_cairo_steps() {
+    //      -> 1 emit event syscall costs 61 cairo steps and 1 range check builtin
+    //      -> 1 range check costs 70
+    //      -> the minimum total cost is `syscall_base_gas_cost`, which is pre-charged by the compiler (atm it is 100 * 100)
+    // 186880 = reported consumed sierra gas
```
Isn't this quite substantial compared to other costs (that were calculated)? I wonder if this isn't caused by not including something in the calculations.

THenry14 replied:
Not necessarily - you must remember this is `l2_gas` for sierra steps. If you were to convert this into "classic" gas values, you'd see that its cost is equivalent to ~31 gas (and the cost before any of my changes is 30, slight increase but this is step heavy so this is somewhat expected).

To be super clear where did I get the numbers from:

l2 gas reported here (so the measure of execution cost) is 1239010 for sierra steps and 1208000 for cairo steps (see assert_gas for `events_contract_cost_sierra_gas` and `events_contract_cost_cairo_steps` respectively).

Converting this to classical L1 gas can be done from simple proportion:

100 L2 gas = 0.0025 L1 Gas
1239010 L2 gas = X 
(so X = 30.98 L1 gas)

1208000 L2 gas = X
(so X = 30.2 L1 gas)

As we can see from the same test on master, the numbers checks out - https://github.com/foundry-rs/starknet-foundry/blob/master/crates/forge/tests/integration/gas.rs#L786 :
`11 + 6 + 13` = 30 (we discard 96, as this is not related to execution - this is actually l1_data_gas, that is now taken by use in a separate resource bound; we can add it here, but we would also have to add it to calculations above; and if we did the numbers would still have to match, as we add the same number in both places).


Hope that clarifies things :+1:

cptartur replied:
The cost `11+6+13=30` is for the test `events_contract_cost` while here we are in the equivalent of `events_cost`.

But here the converted value would be `~24` from the 26 on master which is also in line.

Seems fine for me.

THenry14 committed:
Review changes

THenry14 committed:
Review changes

THenry14 committed:
Update docs regarding sierra-gas tracking (#3056)

Related to #2977

This PR stack aims to add support for new resource tracking (sierra-gas)
and new gas representation (GasVector triplet instead of single l1 gas
value).

-- (https://github.com/foundry-rs/starknet-foundry/pull/3049) Add sierra
gas tracking
-- (https://github.com/foundry-rs/starknet-foundry/pull/3050) Run
already existing tests with CairoSteps set explicitly
-- (https://github.com/foundry-rs/starknet-foundry/pull/3051) Add
support for sierra gas in `--detailed-resources`
-- (https://github.com/foundry-rs/starknet-foundry/pull/3052) Add
support for sierra gas in `--save-trace-data`
-- (https://github.com/foundry-rs/starknet-foundry/pull/3053) Add sierra
version assertion
-- (https://github.com/foundry-rs/starknet-foundry/pull/3054) Introduce
new gas representation
-- (https://github.com/foundry-rs/starknet-foundry/pull/3055) Add gas
tests for sierra-gas tracking
--> (This PR) Update docs with sierra-gas related changes
